samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala

```diff
@@ -409,10 +414,36 @@ object SamzaContainer extends Logging {
 
     info("Setting up metrics reporters.")
 
-    val reporters = MetricsReporterLoader.getMetricsReporters(config, containerName).asScala.toMap ++ customReporters
+    var reporters = MetricsReporterLoader.getMetricsReporters(config, containerName).asScala.toMap ++ customReporters
 
     info("Got metrics reporters: %s" format reporters.keys)
 
+    info("Setting up diagnostics.")
+
+    val diagnosticsManager = if (config.getDiagnosticsEnabled) {
+
+      val diagnosticsReporterName = MetricsConfig.METRICS_SNAPSHOT_REPORTER_NAME_FOR_DIAGNOSTICS
+
+      val publishInterval: Int = config.getMetricsSnapshotReporterInterval(diagnosticsReporterName).getOrElse("60").toInt
+
+      val diagnosticsSystemStream = StreamUtil.getSystemStreamFromNames(new MetricsConfig(config).
+        getMetricsSnapshotReporterStream(diagnosticsReporterName).
+        getOrElse(throw new ConfigException("Missing required config: " +
+          String.format(MetricsConfig.METRICS_SNAPSHOT_REPORTER_STREAM, diagnosticsReporterName))))
+
+      val systemProducer = systemFactories.get(diagnosticsSystemStream.getSystem).get.getProducer(diagnosticsSystemStream.getSystem, config, new MetricsRegistryMap())
+
+      val diagnosticsMetricsReporter = new MetricsSnapshotReporter(systemProducer, diagnosticsSystemStream, publishInterval, jobName, jobId,
+        containerName, Util.getTaskClassVersion(config), Util.getSamzaVersion(), Util.getLocalHost.getHostName,
+        new MetricsSnapshotSerdeV2, config.getMetricsSnapshotReporterBlacklist(diagnosticsReporterName))
+
+      reporters += (diagnosticsReporterName -> diagnosticsMetricsReporter)
+
+      new DiagnosticsManager(jobName, jobId, containerName, execEnvContainerId, Util.getTaskClassVersion(config),
+        Util.getSamzaVersion, Util.getLocalHost.getHostName, diagnosticsSystemStream, systemProducer)
+    }
+    else null
```

Comment:
Use options instead of null. nulls in Scala are unexpected and may cause accidental NPEs in future.